cockatrice/src/pictureloader.cpp

```diff
@@ -0,0 +1,462 @@
+        qSort(sortedSets.begin(), sortedSets.end(), EnabledAndKeyCompareFunctor());
+    }
+}
+
+bool PictureToLoad::nextSet()
+{
+    if (setIndex == sortedSets.size() - 1)
+        return false;
+    ++setIndex;
+    return true;
+}
+
+QString PictureToLoad::getSetName() const
+{
+    if (setIndex < sortedSets.size())
+        return sortedSets[setIndex]->getCorrectedShortName();
+    else
+        return QString("");
+}
+
+CardSet *PictureToLoad::getCurrentSet() const
+{
+    if (setIndex < sortedSets.size())
+        return sortedSets[setIndex];
+    else
+        return 0;
+}
+
+QStringList PictureLoader::md5Blacklist = QStringList()
+    << "db0c48db407a907c16ade38de048a441"; // card back returned by gatherer when card is not found
+
+PictureLoader::PictureLoader()
+    : QObject(0),
+      downloadRunning(false), loadQueueRunning(false)
+{
+    picsPath = settingsCache->getPicsPath();
+    picDownload = settingsCache->getPicDownload();
+
+    connect(this, SIGNAL(startLoadQueue()), this, SLOT(processLoadQueue()), Qt::QueuedConnection);
+    connect(settingsCache, SIGNAL(picsPathChanged()), this, SLOT(picsPathChanged()));
+    connect(settingsCache, SIGNAL(picDownloadChanged()), this, SLOT(picDownloadChanged()));
+
+    networkManager = new QNetworkAccessManager(this);
+    connect(networkManager, SIGNAL(finished(QNetworkReply *)), this, SLOT(picDownloadFinished(QNetworkReply *)));
+
+    pictureLoaderThread = new QThread;
+    pictureLoaderThread->start(QThread::LowPriority);
+    moveToThread(pictureLoaderThread);
+}
+
+PictureLoader::~PictureLoader()
+{
+    pictureLoaderThread->deleteLater();
+}
+
+void PictureLoader::processLoadQueue()
+{
+    if (loadQueueRunning)
+        return;
+
+    loadQueueRunning = true;
+    forever {
+        mutex.lock();
+        if (loadQueue.isEmpty()) {
+            mutex.unlock();
+            loadQueueRunning = false;
+            return;
+        }
+        cardBeingLoaded = loadQueue.takeFirst();
+        mutex.unlock();
+
+        QString setName = cardBeingLoaded.getSetName();
+        QString correctedCardname = cardBeingLoaded.getCard()->getCorrectedName();
+        qDebug() << "Trying to load picture (set: " << setName << " card: " << correctedCardname << ")";
+
+        //The list of paths to the folders in which to search for images
+        QList<QString> picsPaths = QList<QString>() << picsPath + "/CUSTOM/" + correctedCardname;
+
+        if(!setName.isEmpty())
+        {
+            picsPaths   << picsPath + "/" + setName + "/" + correctedCardname
+                        << picsPath + "/downloadedPics/" + setName + "/" + correctedCardname;
+        }
+
+        QImage image;
+        QImageReader imgReader;
+        imgReader.setDecideFormatFromContent(true);
+        bool found = false;
+
+        //Iterates through the list of paths, searching for images with the desired name with any QImageReader-supported extension
+        for (int i = 0; i < picsPaths.length() && !found; i ++) {
```

Comment:
I think it might make this more readable/maintainable to extract the logic out to a `bool cardImageExistsOnDisk` function (or maybe return `QString`/`nullptr` as result). Then in this function you can just `if cardImageExistsOnDisk(card, set)`